src/expr/src/linear.rs

```diff
@@ -533,4 +546,51 @@ impl MapFilterProject {
         // This should probably resemble existing scalar cse.
         unimplemented!()
     }
+
+    /// Removes unused expressions from `self.expressions`.
+    ///
+    /// Expressions are "used" if they are relied upon by any output columns
+    /// or any predicates, even transitively. Any expressions that are not
+    /// relied upon in this way can be discarded.
+    pub fn remove_undemanded(&mut self) {
+        // Determine the demanded expressions to remove irrelevant ones.
+        let mut demand = std::collections::HashSet::new();
+        for (_index, pred) in self.predicates.iter() {
+            demand.extend(pred.support());
+        }
+        demand.extend(self.projection.iter().cloned());
+        for index in (0..self.expressions.len()).rev() {
```

Comment:
minor: I think its worth a comment that we need to do the demand analysis in reverse order here because later expressions can depend on prior expressions

Reply:
That makes sense! Added.